projects/igniteui-angular/src/lib/directives/toggle/toggle.directive.spec.ts

```diff
@@ -212,6 +218,49 @@ describe('IgxToggle', () => {
         expect(toggleElm.classList.contains(HIDDEN_TOGGLER_CLASS)).toBe(true);
     }));
 
+    it('fix for #2798 - Allow canceling of open and close of IgxDropDown through onOpening and onClosing events', fakeAsync(() => {
+        const fixture = TestBed.createComponent(IgxToggleTestComponent);
+        fixture.detectChanges();
+
+        const toggle = fixture.componentInstance.toggle;
+
+        spyOn(toggle.onOpening, 'emit').and.callThrough();
+        spyOn(toggle.onOpened, 'emit').and.callThrough();
+        spyOn(toggle.onClosing, 'emit').and.callThrough();
+        spyOn(toggle.onClosed, 'emit').and.callThrough();
+
+        toggle.onClosing.subscribe((e: CancelableEventArgs) => e.cancel = true);
+
+        toggle.open();
+        fixture.detectChanges();
+
+        tick();
+        expect(toggle.onOpening.emit).toHaveBeenCalledTimes(1);
+
+        tick();
```

Comment:
Do you need to call `tick()` before each expect? I think you only need one when an operation on the control is performed.

Reply:
done :)